File: examples/scala/src/main/scala/mytest/spark/CloudantApp.scala
/*******************************************************************************
* Copyright (c) 2015 IBM Corp.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*******************************************************************************/
package mytest.spark

import org.apache.spark.SparkContext
import org.apache.spark.SparkContext._
import org.apache.spark.SparkConf
import org.apache.spark.sql.SQLContext

/**
 * @author yanglei
 */
object CloudantApp {
  
  def main(args: Array[String]) {

    val conf = new SparkConf().setAppName("Cloudant Spark SQL External Datasource")
    // set protocol to http if needed, default value=https
    // conf.set("cloudant.protocol","http")
    conf.set("cloudant.host","ACCOUNT.cloudant.com")
    conf.set("cloudant.username", "USERNAME")
    conf.set("cloudant.password","PASSWORD")
    val sc = new SparkContext(conf)
    val sqlContext = new SQLContext(sc)
    import sqlContext._

    println("About to test com.cloudant.spark for n_airportcodemapping")
    sqlContext.sql(
        s"""
        |CREATE TEMPORARY TABLE airportTable
        |USING com.cloudant.spark
        |OPTIONS ( database 'n_airportcodemapping')
        """.stripMargin)
    val airportData = sqlContext.sql("SELECT _id, airportName FROM airportTable WHERE _id >= 'CAA' AND _id <= 'GAA' ORDER BY _id")
    airportData.printSchema()
    println(s"Total # of rows in airportData: " + airportData.count())
    airportData.map(t => "code: " + t(0) + ",name:" + t(1)).collect().foreach(println) 


    println("About to test com.cloudant.spark for n_booking") 
    sqlContext.sql(
        s"""
        |CREATE TEMPORARY TABLE bookingTable
        |USING com.cloudant.spark
        |OPTIONS (database 'n_booking')
        """.stripMargin)
    val bookingData = sqlContext.sql("SELECT customerId, dateOfBooking FROM bookingTable WHERE customerId = 'uid0@email.com'")
    bookingData.printSchema()
    bookingData.map(t => "customer: " + t(0) + ", dateOfBooking: " + t(1)).collect().foreach(println) 

    println("About to test com.cloudant.spark for flight with index")
    sqlContext.sql(
        s"""
        |CREATE TEMPORARY TABLE flightTable
        |USING com.cloudant.spark
        |OPTIONS (database 'n_flight', index '_design/view/_search/n_flights')
        """.stripMargin)
    val flightData = sqlContext.sql("SELECT flightSegmentId, scheduledDepartureTime FROM flightTable WHERE flightSegmentId >'AA9' AND flightSegmentId<'AA95'")
    flightData.printSchema()
    flightData.map(t => "flightSegmentId: " + t(0) + ", scheduledDepartureTime: " + t(1)).collect().foreach(println) 
  
  }
}

File: examples/python/CloudantApp.py
#*******************************************************************************
# Copyright (c) 2015 IBM Corp.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#******************************************************************************/
import pprint
from pyspark.sql import SQLContext
from pyspark import SparkContext, SparkConf

conf = SparkConf().setAppName("Cloudant Spark SQL External Datasource in Python")
# define cloudant related configuration:
# set protocol to http if needed, default value=https
# conf.set("cloudant.protocol","http")
conf.set("cloudant.host","ACCOUNT.cloudant.com")
conf.set("cloudant.username", "USERNAME")
conf.set("cloudant.password","PASSWORD")

sc = SparkContext(conf=conf)
sqlContext = SQLContext(sc)

print 'About to test com.cloudant.spark for n_airportcodemapping'
sqlContext.sql("CREATE TEMPORARY TABLE airportTable USING com.cloudant.spark OPTIONS ( database 'n_airportcodemapping')")
      
airportData = sqlContext.sql("SELECT _id, airportName FROM airportTable WHERE _id >= 'CAA' ORDER BY _id")
airportData.printSchema()
print 'Total # of rows in airportData: ' + str(airportData.count())
for code in airportData.collect():
	print code._id

print 'About to test com.cloudant.spark for n_booking'
sqlContext.sql(" CREATE TEMPORARY TABLE bookingTable USING com.cloudant.spark OPTIONS ( database 'n_booking' , jsonstore.rdd.schemaSampleSize '-1' )")
      
bookingData = sqlContext.sql("SELECT customerId, dateOfBooking FROM bookingTable WHERE customerId = 'uid0@email.com'") 
bookingData.printSchema()
for code in bookingData.collect():
	print code.customerId
	print code.dateOfBooking


print 'About to test com.cloudant.spark for n_airportcodemapping'
sqlContext.sql(" CREATE TEMPORARY TABLE airportTable1 USING com.cloudant.spark OPTIONS ( database 'n_airportcodemapping')")
      
airportData = sqlContext.sql("SELECT _id, airportName FROM airportTable1 WHERE _id >= 'CAA' AND _id <= 'GAA' ORDER BY _id")
airportData.printSchema()
print 'Total # of rows in airportData: ' + str(airportData.count())
for code in airportData.collect():
	print code._id

print 'About to test com.cloudant.spark for n_booking'
sqlContext.sql(" CREATE TEMPORARY TABLE bookingTable1 USING com.cloudant.spark OPTIONS ( database 'n_booking')")
      
bookingData = sqlContext.sql("SELECT customerId, dateOfBooking FROM bookingTable1 WHERE customerId = 'uid0@email.com'")
bookingData.printSchema()
for code in bookingData.collect():
	print code.customerId
	print code.dateOfBooking

print 'About to test com.cloudant.spark for flight with index'
sqlContext.sql(" CREATE TEMPORARY TABLE flightTable1 USING com.cloudant.spark OPTIONS ( database 'n_flight', index '_design/view/_search/n_flights')")
      
flightData = sqlContext.sql("SELECT flightSegmentId, scheduledDepartureTime FROM flightTable1 WHERE flightSegmentId >'AA9' AND flightSegmentId<'AA95'")
flightData.printSchema()
for code in flightData.collect():
	print 'Flight {0} on {1}'.format(code.flightSegmentId, code.scheduledDepartureTime)
		

print 'About to test com.cloudant.spark for n_airportcodemapping'
sqlContext.sql(" CREATE TEMPORARY TABLE airportTable2 USING com.cloudant.spark OPTIONS ( database 'n_airportcodemapping')")
      
airportData = sqlContext.sql("SELECT _id, airportName FROM airportTable2 WHERE _id >= 'CAA' AND _id <= 'GAA' ORDER BY _id")
airportData.printSchema()
for code in airportData.collect():
	print code._id

print 'About to test com.cloudant.spark for n_booking'
sqlContext.sql(" CREATE TEMPORARY TABLE bookingTable2 USING com.cloudant.spark OPTIONS ( database 'n_booking')")
      
bookingData = sqlContext.sql("SELECT customerId, dateOfBooking FROM bookingTable2 WHERE customerId = 'uid0@email.com'")
bookingData.printSchema()
for code in bookingData.collect():
	print 'Booking for {0} on {1}'.format(code.customerId,code.dateOfBooking)

print 'About to test com.cloudant.spark for flight with index'
sqlContext.sql(" CREATE TEMPORARY TABLE flightTable2 USING com.cloudant.spark OPTIONS ( database 'n_flight', index '_design/view/_search/n_flights')")
      
flightData = sqlContext.sql("SELECT flightSegmentId, scheduledDepartureTime FROM flightTable2 WHERE flightSegmentId >'AA9' AND flightSegmentId<'AA95'")
flightData.printSchema()
for code in flightData.collect():
	print 'Flight {0} on {1}'.format(code.flightSegmentId, code.scheduledDepartureTime)

sc.stop()
